lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,468 @@
+	}
+
+	var (
+		redisConn      redis.UniversalClient
+		connectionAddr = fmt.Sprintf("%s:%s", connectionOptions.address, connectionOptions.port)
+	)
+
+	// TODO(jakub): Use system CA bundle if connecting to AWS.
+	// TODO(jakub): Investigate Redis Sentinel.
+	if connectionOptions.cluster {
+		redisConn = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs:     []string{connectionAddr},
+			TLSConfig: tlsConfig,
+		})
+	} else {
+		redisConn = redis.NewClient(&redis.Options{
+			Addr:      connectionAddr,
+			TLSConfig: tlsConfig,
+		})
+	}
+	defer redisConn.Close()
+
+	e.Log.Debug("created a new Redis client, sending ping to test the connection")
+
+	// TODO(jakub): Currently Teleport supports only RESP2 as RESP3 is not supported by go-redis.
+	// When migration to RESP3 protocol this PING should be removed in favor of "HELLO 3" cmd.
+	// https://github.com/antirez/RESP3/blob/master/spec.md#the-hello-command-and-connection-handshake
+	pingResp := redisConn.Ping(context.Background())
+	if pingResp.Err() != nil {
+		return trace.Wrap(pingResp.Err())
+	}
+
+	e.Log.Debug("Redis server responded to ping message")
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	if err := e.process(ctx, redisConn); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// readClientCmd reads commands from connected Redis client.
+func (e *Engine) readClientCmd(ctx context.Context) (*redis.Cmd, error) {
+	cmd := &redis.Cmd{}
+	if err := cmd.ReadReply(e.clientReader); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	val, ok := cmd.Val().([]interface{})
+	if !ok {
+		return nil, trace.BadParameter("failed to cast Redis value to a slice")
+	}
+
+	return redis.NewCmd(ctx, val...), nil
+}
+
+// processCmd processes commands received from connected client. Most commands are just passed to Redis instance,
+// but some require special actions:
+//  * Redis 7.0+ commands are rejected as at the moment of writing Redis 7.0 hasn't been released and go-redis doesn't support it.
+//  * RESP3 commands are rejected as Teleport currently doesn't support this version of protocol.
+//  * Subscribe related commands created a new DB connection as they change Redis request-response model to Pub/Sub.
+func (e *Engine) processCmd(ctx context.Context, redisClient redis.UniversalClient, cmd *redis.Cmd) error {
+	// go-redis always returns lower-case commands.
+	switch cmd.Name() {
+	case "hello":
+		return redis.RedisError("RESP3 is not supported")
+	case "auth":
+		return e.processAuth(ctx, redisClient, cmd)
+	case "subscribe":
+		return e.subscribeCmd(ctx, redisClient.Subscribe, cmd)
+	case "psubscribe":
+		return e.subscribeCmd(ctx, redisClient.PSubscribe, cmd)
+	case "punsubscribe":
+		// TODO(jakub): go-redis doesn't expose any API for this command. Investigate alternative options.
+		return errors.New("PUNSUBSCRIBE is not supported by Teleport")
```

Comment:
```suggestion
		return trace.NotImplemented("PUNSUBSCRIBE is not supported by Teleport")
```

Reply:
The reason why I'm not using trace package here is that I want plain errors as they are send to connected client https://github.com/gravitational/teleport/pull/10053/files#diff-937488b4e06d9efb0aa05688ddfb2dc375598f2601043b4acaad10d7219af517R395
But now when I think about it I think that:
1. I can use `redis.Error` to make it more explicit
2. Add additional handlers for `trace` errors. That would also handle errors types better, as now all errors are sent as generic errors.